### PR TITLE
*: kill one's own connection doesn't require SUPER privilege

### PR DESCRIPTION
**Pull request by tiancaiamao**
## What have you changed? (mandatory)

`kill tidb connID`, if the user is the owner of that connection, there
is no need to check the SUPER privilege.
SessionManager interface is slightly modified.

## What are the type of the changes (mandatory)?

- Bug fix (non-breaking change which fixes an issue)

## How has this PR been tested (mandatory)?

It's somehow difficult to test this change. I check it manually.

```
mysql> select current_user();
+----------------+
| current_user() |
+----------------+
| kill@127.0.0.1 |
+----------------+
1 row in set (0.00 sec)

mysql> show processlist;
+------+------+-----------+------+---------+------+-------+------------------+------+
| Id   | User | Host      | db   | Command | Time | State | Info             | Mem  |
+------+------+-----------+------+---------+------+-------+------------------+------+
|    3 | kill | 127.0.0.1 |      | Query   |    2 | 2     | select sleep(40) |    0 |
|    4 | kill | 127.0.0.1 |      | Query   |    0 | 2     | show processlist |    0 |
|    2 | root | 127.0.0.1 |      | Query   |    0 | 2     |                  |    0 |
+------+------+-----------+------+---------+------+-------+------------------+------+
3 rows in set (0.00 sec)

mysql> kill tidb 2;                 // Kill other's connection without SUPER privilege would fail.
ERROR 1105 (HY000): privilege check fail
mysql> kill tidb 3;                 // Kill the user's own connection.
Query OK, 0 rows affected (0.00 sec)
```

PTAL @jackysp

### CODE REVIEW

File: plan/planbuilder.go [hunk 1]
```diff
@@ -783,8 +783,21 @@ func (b *planBuilder) buildSimple(node ast.StmtNode) Plan {
...
 	case *ast.GrantStmt:
 		b.visitInfo = collectVisitInfoFromGrantStmt(b.visitInfo, raw)
-	case *ast.SetPwdStmt, *ast.RevokeStmt, *ast.KillStmt:
+	case *ast.SetPwdStmt, *ast.RevokeStmt:
 		b.visitInfo = appendVisitInfo(b.visitInfo, mysql.SuperPriv, "", "", "")
+	case *ast.KillStmt:
+		// If you have the SUPER privilege, you can kill all threads and statements.
+		// Otherwise, you can kill only your own threads and statements.
+		sm := b.ctx.GetSessionManager()
+		if sm != nil {
```
Comment (shenli) [inline plan/planbuilder.go]: Could the `sm` be nil?

File: util/processinfo.go [hunk 2]
```diff
@@ -33,6 +33,6 @@ type ProcessInfo struct {
 // SessionManager is an interface for session manage. Show processlist and
 // kill statement rely on this interface.
 type SessionManager interface {
-	ShowProcessList() []ProcessInfo
+	ShowProcessList() map[uint64]ProcessInfo
```
Comment (shenli) [inline util/processinfo.go]: Add comment for the return value.

File: plan/planbuilder.go [hunk 3]
```diff
@@ -783,8 +783,21 @@ func (b *planBuilder) buildSimple(node ast.StmtNode) Plan {
...
+			processList := sm.ShowProcessList()
```
Comment (shenli) [inline plan/planbuilder.go]: Could you add an interface for SessionManger like GetProcessByConnID(connID) ?

**Comment by shenli** [comment]
@tiancaiamao Please fix the CI.

File: plan/planbuilder.go (hunk 1 above)
Comment (tiancaiamao) [reply plan/planbuilder.go]: In the test code

File: plan/planbuilder.go (hunk 3 above)
Comment (tiancaiamao) [reply plan/planbuilder.go]: `GetProcessByConnID` make `SessionManager` more complex.
`GetProcessByConnID` can be implemented using `ShowProcessList` which is already in `SessionManager`
I assume `Kill` will not be too frequency so this choice would not hurt the performance.
@shenli

**Comment by tiancaiamao** [comment]
PTAL @shenli @zimulala

File: plan/planbuilder.go (hunk 3 above)
Comment (shenli) [reply plan/planbuilder.go]: So the interface `ShowProcessList` return a map instead of a list.

Comment (tiancaiamao) [reply plan/planbuilder.go]: Yes.

**Pull request merged**